recipes/librdkafka/all/conanfile.py

```diff
@@ -117,6 +114,16 @@ def package_info(self):
         self.cpp_info.components["rdkafka"].names["cmake_find_package_multi"] = "rdkafka"
         self.cpp_info.components["rdkafka"].names["pkg_config"] = "rdkafka"
         self.cpp_info.components["rdkafka"].libs = ["rdkafka"]
+        if self.options.zlib:
```

Comment:
This new “feature” sure requires a lot of boilerplate.

Reply:
@jgsogo what do you think

Reply:
What do you mean? I think I need a longer explanation. Thanks!

Reply:
Sorry - I mean this new section:
```
        if self.options.zlib:
            self.cpp_info.components["rdkafka"].requires.append("zlib::zlib")
        if self.options.zstd:
            self.cpp_info.components["rdkafka"].requires.append("zstd::zstd")
        if self.options.ssl:
            self.cpp_info.components["rdkafka"].requires.append("openssl::openssl")
        if self.options.sasl and self.settings.os != "Windows":
            self.cpp_info.components["rdkafka"].requires.append("cyrus-sasl::cyrus-sasl")
```
It's required otherwise an exception is raised.   Will every recipe need to list each of its requirements also as `components[x].requires`?    It seems like boilerplate to me.  Would it be possible to auto-add them?  Or have a helper function in `package_info()` that does?

Reply:
I don't think conan can determine this information.

Reply:
We were requested to list explicitly the requirements that are consumed by each library, the idea was originated related to the components, but now (1.30) it is a feature for packages without components too: https://github.com/conan-io/conan/pull/7644

Probably it is impossible to infer something like this and auto-add them, it would be against the feature, but if needed we can add a `tool` to make things easier... nevertheless I tend to think that recipes are written once, so this is something we will suffer while adding components, but afterward we will almost forget about it.

Reply:
What about a helper function like `self.add_component_requires("rdkafka")`?

Reply:
It needs the name of the option and the name of component to add too, something like (quick draft):

```python
def add_component_requires(conanfile, component, option, require_component=None):
    require_component = require_component or f"{option}:{option}"
    if conanfile.options.get_safe(option):
        conanfile.cpp_info.compoentns[component].requires.append(require_component)
```

and it will be used like:

```python
tools.add_component_requires(self, 'rdkafka', 'ssl', 'openssl::openssl')
tools.add_component_requires(self, 'rdkafka', 'zlib')
```

instead of:

```python
if self.options.zlib:
    self.cpp_info.components["rdkafka"].requires.append("zlib::zlib")
if self.options.ssl:
    self.cpp_info.components["rdkafka"].requires.append("openssl::openssl")
```

I understand it is shorter, but it is less explicit in return for one line. I'm not sure if it is worth hiding the `if` and the full component-related line.

Reply:
I agree - if you need to make n calls to a helper for n requires then it's not worth it.  However, the package recipes are already known, and in this example all of the components requires are of the form
```
self.cpp_info.components[comp].requires.append("requires::requires")
```
Where `requires` is the name of a package requires.  So it seems like a call like:
```
self.add_component_requires("rdkafka")
```
Could be implemented like (pseudocode):
```
def add_component_requires(component):
    for dependency in package_requires:
       self.cpp_info.components[component].requires.append(f"{dependency.name}::{dependency.name}")
```
So you only need to call one function to add all component requires.

Reply:
A component can require some dependency, not all. And a component can be linked to a component of a dependency.
With your method, you introduce useless links between components and dependency.

For example, my package A have two components B and C. B is linked with a library M and C is linked with a component P from a package O (which contains two or more components). 
With your method, B will be linked yo M and all components of O (although it does not use any of them). And C will be linked to M and all unused components of O.

The components have been created to handle this situation and to avoid unnecessary links. Adding a boilerplate as you propose would make the components useless.

Reply:
You would only call it in the common case (which I assume this package is a representative).  If you have more specialized need you would do things manually.